File: Java/Problem_017.java
public class Main {
    static final int[] ones = {0,3,3,5,4,4,3,5,5,4};

    static final int[] teens = new int[20];
    static final int[] tens = new int[10];

    static final int HUNDRED = 7, AND = 3, THOUSAND = 8;

    static {
        teens[10]=3; teens[11]=6; teens[12]=6; teens[13]=8; teens[14]=8;
        teens[15]=7; teens[16]=7; teens[17]=9; teens[18]=8; teens[19]=8;

        tens[2]=6; tens[3]=6; tens[4]=5; tens[5]=5; tens[6]=5; tens[7]=7; tens[8]=6; tens[9]=6;
    }

    static int letters(int n) {
        if (n == 1000) return ones[1] + THOUSAND;

        int total = 0;

        if (n >= 100) {
            int h = n / 100;
            total += ones[h] + HUNDRED;
            if (n % 100 != 0) total += AND;
            n %= 100;
        }

        if (n >= 20) {
            total += tens[n / 10];
            total += ones[n % 10];
        } else if (n >= 10) {
            total += teens[n];
        } else {
            total += ones[n];
        }

        return total;
    }

    public static void main(String[] args) {
        long sum = 0;
        for (int i = 1; i <= 1000; i++) sum += letters(i);
        System.out.println(sum); // 21124
    }
}

File: Python/Problem_017.py
def solve() -> int:
    ones = {
        0: 0,
        1: 3,  # one
        2: 3,  # two
        3: 5,  # three
        4: 4,  # four
        5: 4,  # five
        6: 3,  # six
        7: 5,  # seven
        8: 5,  # eight
        9: 4,  # nine
    }

    teens = {
        10: 3,  # ten
        11: 6,  # eleven
        12: 6,  # twelve
        13: 8,  # thirteen
        14: 8,  # fourteen
        15: 7,  # fifteen
        16: 7,  # sixteen
        17: 9,  # seventeen
        18: 8,  # eighteen
        19: 8,  # nineteen
    }

    tens = {
        2: 6,  # twenty
        3: 6,  # thirty
        4: 5,  # forty
        5: 5,  # fifty
        6: 5,  # sixty
        7: 7,  # seventy
        8: 6,  # eighty
        9: 6,  # ninety
    }

    HUNDRED = 7   # "hundred"
    AND = 3       # "and"
    THOUSAND = 8  # "thousand"

    def letters(n: int) -> int:
        if n == 1000:
            return ones[1] + THOUSAND

        total = 0

        if n >= 100:
            h = n // 100
            total += ones[h] + HUNDRED
            if n % 100 != 0:
                total += AND
            n %= 100

        if n >= 20:
            total += tens[n // 10]
            total += ones[n % 10]
        elif n >= 10:
            total += teens[n]
        else:
            total += ones[n]

        return total

    return sum(letters(i) for i in range(1, 1001))


if __name__ == "__main__":
    print(solve())  # 21124
